Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php

```diff
@@ -146,6 +146,20 @@ public function registerNodeChange(NodeInterface $node, Workspace $targetWorkspa
      */
     protected function registerAllTagsToFlushForNodeInWorkspace(NodeInterface $node, Workspace $workspace): void
     {
+        // Ensure that we're dealing with the variant of the given node that actually
+        // lives in the given workspace
+        if ($node->getWorkspace()->getName() !== $workspace->getName()) {
```

Comment:
@grebaldi do you think this could be a performance problem when working with a large number >1000 nodes?
Like publishing a whole workspace and for some reason creating a large number of contexts?

Reply:
Good question :)

Context creation itself doesn't strike me as very expensive, so I guess that would not be a problem. `$workspaceContext->getNodeByIdentifier` however will be quite expensive given an n=1000 iteration. It cannot be avoided though (IIRC).

Reply:
Ah i see each `Context` has its own `firstLevelNodeCache`, that makes the caching of course workspace ware, which makes sense. And as we create the context just at that very second, the cache is completely empty so indeed we will query the content repository.

Reply:
I dont fully understand this yet, but if we only want to ensure that `$node->getWorkspace()->getName() === $workspace->getName()`
we _can_ cheat by building our own node:

```php
$workspaceContext = $this->contextFactory->create(
    array_merge(
        $node->getContext()->getProperties(),
        ['workspaceName' => $workspace->getName()]
    )
);
$correctNode = $this->nodeFactory->createFromNodeData($node->getNodeData(), $workspaceContext);
```

but in case that is not our sole goal but we want to really ask the database, so the `if ($node === null) return;` will be used, that cheating will of course not work.